### DISCUSSION TITLE
Is it possible to configure autoHighlight in CompositeLayer with complex data?

### DISCUSSION THREAD

**Discussion by Dv-Andrew** [visgl/deck.gl#7591] [Q&A] [answered]
I'm trying to combine few `CompositeLayers` to create a layer that will be able to render a lines using the `PathLayer` and points at the nodes of this path.
But deck.gl create similar `indexes` in the main layer and its sublayers. This leads to a behavior, when cursor hover one of the entities, another one with an identical index is highlighted too:

https://user-images.githubusercontent.com/37058479/213728673-c2c9aa32-846d-4980-8831-c2d6b7b25d65.mov

To resolve this problem, I can split the lines data and use two different layers to render them.
But it will be handy if I can use only one `CompositeLayer` to work with such lines and processing all their data inside the layer.
The main question is — will it be possible to do this for my dataset? Maybe I'm missing something and the layer needs to be improved?

Example of data:
```
{
  "id": "some-line-id",
  "type": "LINE",
  "points": [
    {
      "id": "some-point-id",
      "coordinates": {
        "easting": 6.945529685615371,
        "northing": 51.49951462325985
      },
      // some metadata
    },
    ...
  ],
  "path": [
    [
      6.945529685615371,
      51.49951462325985
    ],
    ...
  ]
}

```

Example of layer:
```
// Composite PointsLayer replaced with ScatterplotLayer for simplicity

export const defaultProps = {
  id: 'lines-layer',
  data: [],
  pickable: true,
  autoHighlight: true
};


export class LinesLayer extends CompositeLayer {
  initializeState() {
    this.state = {
      path: [],
      points: []
    };
  }

  updateState(params) {
    const { props, changeFlags } = params;

    if (changeFlags.dataChanged) {
      const newPath = [];
      const newPoints = [];

      props.data.forEach((object, index) => {
        newPath.push(this.getSubLayerRow({ path: object.path }, object, index));

        for (const point of object.points) {
          newPoints.push(this.getSubLayerRow({ point }, object, index));
        }
      });

      this.setState({ path: newPath, points: newPoints });
    }
  }

  renderLayers() {
    const {
      pickable,
      autoHighlight,
      onClick
    } = this.props;

    const { path, points } = this.state;

    return [
      new PathLayer(
        this.getSubLayerProps({
          id: 'path'
        }),
          {
            data: path,
            getPath: ({ path }) => path,
            onClick,
            pickable,
            autoHighlight,
            // other props

            updateTriggers: this.props.updateTriggers
          }
      ),
      new ScatterplotLayer(
        this.getSubLayerProps({
          id: 'points'
        }),
        {
          data: points,
          onClick,
          getPosition: ({ point }) => [
            point.coordinates.easting,
            point.coordinates.northing
          ],
          // other props

          updateTriggers: this.props.updateTriggers
        }
      )
    ];
  }
}

LinesLayer.layerName = 'LinesLayer';
LinesLayer.defaultProps = defaultProps;

```

**Comment by Pessimistress** [comment, marked as answer]
These is indeed a (undocumented) lifecycle method on each layer called `updateAutoHighlight` (which calls a protected implementation `_updateAutoHighlight`):

https://github.com/visgl/deck.gl/blob/e923ad9261b366d223c537f6641fef85524c4498/modules/core/src/lib/layer.ts#L1107-L1111

By default `CompositeLayer` just passes the picking info down to all sublayers:

https://github.com/visgl/deck.gl/blob/e923ad9261b366d223c537f6641fef85524c4498/modules/core/src/lib/composite-layer.ts#L250-L254

You can see how `GeoJsonLayer` selectively highlight some sublayers:

https://github.com/visgl/deck.gl/blob/e923ad9261b366d223c537f6641fef85524c4498/modules/layers/src/geojson-layer/geojson-layer.ts#L397-L417

It is currently undocumented because we still have debates on how the process should be exposed. However it is not going to change in the 8.x timeframe.

**Reply by Dv-Andrew** [reply]
Thank you very much!
It looks like exactly what I was looking for.

**Reply by ted341** [reply]
@Pessimistress - does that mean we need to split the data into separate layers to avoid double-highlighting? And is there any update on selectively highlighting the sublayers from the CompositeLayer class?
